File: bindings/python/sample_ppc.py
#!/usr/bin/env python
# Sample code for MIPS of Unicorn. Nguyen Anh Quynh <aquynh@gmail.com>
# Python sample ported by Loi Anh Tuan <loianhtuan@gmail.com>

from __future__ import print_function
from unicorn import *
from unicorn.ppc_const import *
from keystone import *
import struct
import sys

ks = Ks(KS_ARCH_PPC,KS_MODE_PPC32 + KS_MODE_BIG_ENDIAN)

def asm(instruction):
    code = ks.asm(instruction)[0]
    if sys.version_info[0] >= 3:
        return bytes(code)
    else:
        return str(bytearray(code))


# callback for tracing basic blocks
def hook_block(uc, address, size, user_data):
    print(">>> Tracing basic block at 0x%x, block size = 0x%x" %(address, size))

# callback for tracing instructions
def hook_code(uc, address, size, user_data):
    print(">>> Tracing instruction at 0x%x, instruction size = 0x%x" %(address, size))

# callback for cpu interruption
def hook_intr(uc, intno, user_data):
    print("INTR : %d" % intno)

# code to be emulated
PPC_CODE_EB  = asm("lwz %r1,0(%r1)")
# memory address where emulation starts
ADDRESS      = 0x10000
DATA = 0x800000


def main():
    # Initialize emulator in POWERPC
    mu = Uc(UC_ARCH_PPC, UC_MODE_BIG_ENDIAN | UC_MODE_32)
    # map 2MB memory for this emulation
    mu.mem_map(ADDRESS, 2 * 1024 * 1024)
    # tracing all basic blocks with customized callback
    mu.hook_add(UC_HOOK_BLOCK, hook_block)
    # tracing all instructions with customized callback
    mu.hook_add(UC_HOOK_CODE, hook_code)
    # write machine code to be emulated to memory
    mu.mem_write(ADDRESS, PPC_CODE_EB)

    mu.mem_map(DATA,4096)
    mu.mem_write(DATA,struct.pack(">I",4))

    mu.reg_write(UC_PPC_REG_GPR_1,DATA)

    # mu.hook_add(UC_HOOK_INTR,hook_intr)

    try:
        mu.emu_start(ADDRESS, ADDRESS + len(PPC_CODE_EB))
    except UcError as e:
        print("ERROR: %s" % e)

    print("GPR1:%08.8x" % mu.reg_read(UC_PPC_REG_GPR_1))

if __name__=='__main__':
    main()

File: bindings/ruby/sample_ppc.rb
#!/usr/bin/env ruby
# Sample code for MIPS of Unicorn. Nguyen Anh Quynh <aquynh@gmail.com>
# Ruby sample ported by Sascha Schirra <sashs82@gmail.com>
require 'unicorn'
require 'unicorn/ppc_const'

include Unicorn

# code to be emulated
PPC_CODE_EB = "\x39\x20\x00\x04" + # li        r9, 4
              "\x91\x3F\x00\x08" + # stw       r9, 8(r31)
              "\x39\x20\x00\x05" + # li        r9, 5
              "\x91\x3F\x00\x0C" + # stw       r9, 0xC(r31)
              "\x81\x5F\x00\x08" + # lwz       r10, 8(r31)
              "\x81\x3F\x00\x0C" + # lwz       r9, 0xC(r31)
              "\x7D\x2A\x4A\x14";  # add       r9, r10, r9" # ori $at, $at, 0x3456;

# memory address where emulation starts
ADDRESS      = 0x10000


# callback for tracing basic blocks
$hook_block = Proc.new do |uc, address, size, user_data|
    puts(">>> Tracing basic block at 0x%x, block size = 0x%x" % [address, size])
end


# callback for tracing instructions
$hook_code = Proc.new do |uc, address, size, user_data|
    puts(">>> Tracing instruction at 0x%x, instruction size = %u" % [address, size])
end

# Test PPC EB
def test_ppc_eb()
    puts("Emulate MIPS code (big-endian)")
    begin
        # Initialize emulator in MIPS32 + EB mode
        mu = Uc.new UC_ARCH_PPC, UC_MODE_PPC32 + UC_MODE_BIG_ENDIAN

        # map 2MB memory for this emulation
        mu.mem_map(ADDRESS, 2 * 1024 * 1024)

        # write machine code to be emulated to memory
        mu.mem_write(ADDRESS, PPC_CODE_EB)

        # initialize machine registers
        mu.reg_write(UC_PPC_REG_GPR_31, ADDRESS)

        # tracing all basic blocks with customized callback
        mu.hook_add(UC_HOOK_BLOCK, $hook_block)

        # tracing all instructions with customized callback
        mu.hook_add(UC_HOOK_CODE, $hook_code)

        # emulate machine code in infinite time
        mu.emu_start(ADDRESS, ADDRESS + PPC_CODE_EB.bytesize)

        # now puts out some registers
        puts(">>> Emulation done. Below is the CPU context")

        r9 = mu.reg_read(UC_PPC_REG_GPR_9)
        puts(">>> r1 = 0x%x" % r9)
        if r9 != 9
          puts("INVALID RETURN VALUE")
        end

    rescue UcError => e
        puts("ERROR: %s" % e)
    end
end


test_ppc_eb()
